src/dragonball/src/dbs_pci/src/vfio.rs

```diff
@@ -0,0 +1,1918 @@
+            vfio_container,
+        )?;
+
+        state.probe_regions(bus)?;
+        state.parse_capabilities();
+        state.parse_legacy_irq();
+        state.fixup_msix_region();
+
+        Ok(VfioPciDevice {
+            id,
+            vm_fd,
+            state: Mutex::new(state),
+        })
+    }
+
+    /// Get resource requirements of the VFIO PCI device.
+    pub fn get_resource_requirements(&self, requests: &mut Vec<ResourceConstraint>) {
+        let state = self.state();
+        if !state.sys_constraints.is_empty() {
+            let mut constraints = state.sys_constraints.clone();
+            requests.append(&mut constraints);
+        }
+    }
+
+    /// Because when routing MSI, KVM needs to know the device id of the device corresponding to the MSI.
+    /// Therefore, when VfioPciDevice is activated, we will record its device id in its irq_manager
+    /// structure for routing MSI.
+    #[cfg(target_arch = "aarch64")]
+    fn set_device_id(&self, state: &mut VfioPciDeviceState<C>) {
+        let device_id = Some(self.id as u32);
+        state
+            .interrupt
+            .irq_manager
+            .as_mut()
+            .unwrap()
+            .set_device_id(device_id);
+    }
+
+    pub fn activate(&self, device: Weak<dyn DeviceIo>, resources: DeviceResources) -> Result<()> {
+        let mut state = self.state();
+
+        if resources.len() == 0 {
+            return Err(VfioPciError::InvalidResources);
+        }
+
+        let mut all_kvm_slot = resources.get_kvm_mem_slots();
+
+        let mut reg_idx = 0;
+        while reg_idx < state.regions.len() {
+            if state.regions[reg_idx].mappable() {
+                state.regions[reg_idx].mmaps[0].slot = all_kvm_slot.pop().unwrap();
+                if state.regions[reg_idx].mmaps.len() == 2 {
+                    state.regions[reg_idx].mmaps[1].slot = all_kvm_slot.pop().unwrap();
+                }
+            }
+            state.add_bar_for_region(reg_idx)?;
+            reg_idx += 1;
+        }
+
+        if let Some(base) = resources.get_legacy_irq() {
+            assert!(base < 0xffu32);
+            state
+                .configuration
+                .set_irq(base as u8, PciInterruptPin::IntA);
+            state.interrupt.add_legacy_irq_resource(base);
+        }
+
+        if let Some((base, size)) = resources.get_pci_msix_irqs() {
+            state.interrupt.add_msi_irq_resource(base, size);
+        }
+
+        let irq_manager = state.context.upgrade().unwrap().get_interrupt_manager();
+        state.interrupt.initialize(irq_manager)?;
+        #[cfg(target_arch = "aarch64")]
+        self.set_device_id(&mut state);
+        state.device = Some(device);
+        if let Err(e) = state.register_regions(&self.vm_fd) {
+            error!(
+                "{} register regions failed. error: {}",
+                state.vfio_path(),
+                e
+            );
+            let _ = state.unregister_regions(&self.vm_fd).map_err(|_e| {
```

Comment:
don't discard internal errors, and explain why we need to panic?

Reply:
error print out in the panic information and also comment added why we need to panic here.